tests/lifecycle/test_lifecycle_adapters_end_to_end_task_based.py

```diff
@@ -0,0 +1,301 @@
+from collections import Counter
+from types import ModuleType
+from typing import Any, Dict, List, Optional
+
+import pytest
+
+from hamilton import ad_hoc_utils, driver, node
+from hamilton.execution.executors import SynchronousLocalTaskExecutor
+from hamilton.graph import FunctionGraph
+from hamilton.htypes import Collect, Parallelizable
+from hamilton.lifecycle.base import (
+    BaseDoNodeExecute,
+    BasePostGraphConstruct,
+    BasePostGraphExecute,
+    BasePostNodeExecute,
+    BasePostTaskExecute,
+    BasePreDoAnythingHook,
+    BasePreGraphExecute,
+    BasePreNodeExecute,
+    BasePreTaskExecute,
+)
+from hamilton.node import Node
+
+from .lifecycle_adapters_for_testing import (
+    ExtendToTrackCalls,
+    SentinelException,
+    TrackingDoNodeExecuteHook,
+    TrackingPostNodeExecuteHook,
+    TrackingPostTaskExecuteHook,
+    TrackingPreNodeExecuteHook,
+)
+
+
+def _sample_driver(*lifecycle_adapters):
+    def n_iters(n_iters_input: int) -> int:
+        return n_iters_input
+
+    def parallel_over(n_iters: int) -> Parallelizable[int]:
+        for i in range(n_iters):
+            yield i
+
+    def processed(parallel_over: int) -> int:
+        return parallel_over**2
+
+    def more_processed(processed: int, broken: bool = False) -> int:
+        if broken:
+            raise SentinelException()
+        return processed**2
+
+    def collect(more_processed: Collect[int]) -> int:
+        return sum(more_processed)
+
+    def output(collect: int) -> int:
+        return collect
+
+    mod = ad_hoc_utils.create_temporary_module(
+        n_iters, parallel_over, processed, more_processed, collect, output
+    )
+    return (
+        driver.Builder()
+        .with_modules(mod)
+        .with_adapters(*lifecycle_adapters)
+        .enable_dynamic_execution(allow_experimental_mode=True)
+        .with_remote_executor(SynchronousLocalTaskExecutor())
+        .build()
+    )
+
+
+def test_individual_pre_node_execute_hook_task_based():
+    hook_name = "pre_node_execute"
+    hook = TrackingPreNodeExecuteHook(name=hook_name)
+    dr = _sample_driver(hook)
+    dr.execute(["output"], inputs={"n_iters_input": 5})
+    relevant_calls = [item for item in hook.calls if item.name == hook_name]
+    assert len(relevant_calls) == 14
+    nodes_executed = Counter([item.bound_kwargs["node_"].name for item in relevant_calls])
+    assert nodes_executed == {
+        "parallel_over": 1,
+        "n_iters": 1,
+        "processed": 5,
+        "more_processed": 5,
+        "collect": 1,
+        "output": 1,
+    }
+    run_ids = {item.bound_kwargs["run_id"] for item in relevant_calls}
+    (run_id,) = run_ids
+    assert len(run_ids) == 1
+    assert len(run_id) > len("TODO -- add a run ID")  # I'd hope it has more entropy than this...
```

Comment:
what are you testing for here?

Reply:
Its a UUID, it should have more chars than the p;laceholder I had before :) i'll change